### PR TITLE
Add auto-download support for Deluge

### PR DESCRIPTION
**Pull request by qwe2**
I have been using this for quite a while in another format (as a stand-alone app that calls `deluge-console.exe` with the arguments coming from Taiga) to make my life easier and enable Deluge to automatically download torrents just the same as uTorrent can. I decided somebody else might find this useful so I decided to add it to Taiga itself (it's also more convinient to use for me this way).

For this I have refactored `Aggregator::HandleFeedDownloadOpen` a bit to work not only for uTorrent but for a given `TorrentClient` instance. I added implementation for Deluge and uTorrent but it would be quite straightforward to add additional torrent clients as well.

I'm not sure my code is up to standards but I was trying to keep myself to the other parts of the project and the best practices I know of.

Also a caveat for this to work is that Deluge must be running in non-classic mode (which it does not do by default). I am not aware of any way to add torrents with an associated path.

### CODE REVIEW

**Comment by erengy** [comment]
Hey there @qwe2, and thanks for the pull request!

The code is up to standards for the most part, although there are a couple of important issues that I'm going to mention via inline comments.

The main issue, however, is that the solution is over-engineered. Pretty much the entirety of `torrent_client.cpp` and `torrent_client.h` is unnecessary for such a small task. You could get the same functionality with just about a dozen new lines, instead of all those objects and smart pointers and stuff. Besides, it's unlikely that we're going to support any more clients in near future.

I went ahead and created a simpler alternative at the [`deluge` branch](https://github.com/erengy/taiga/compare/master...deluge). Could you check it out and see if it works?

File: src/base/file.cpp [hunk 1]
```diff
@@ -132,7 +132,7 @@ bool Execute(const std::wstring& path, const std::wstring& parameters) {
     return ExecuteFile(path, parameters);
 
   auto value = ShellExecute(nullptr, L"open", path.c_str(), parameters.c_str(),
-                            nullptr, SW_SHOWNORMAL);
+                            nullptr, SW_HIDE);
```
Comment (erengy) [inline src/base/file.cpp]: Note that you changed the default behavior of a common function that is used throughout the application. The same function is used to run media players, open anime folders, etc. While this change makes `deluge-console.exe` run seamlessly, it ends up creating multiple bugs.

File: src/track/feed_aggregator.cpp [hunk 2]
```diff
@@ -337,14 +335,23 @@ void Aggregator::HandleFeedDownloadOpen(FeedItem& feed_item,
...
+    client->DownloadTorrent(app_path, download_path, file);
   }
+  else {
+    // Default behavior if no application was found
+    std::wstring parameters;
 
-  parameters += L"\"" + file + L"\"";
-  Execute(app_path, parameters);
+    if (!download_path.empty())
+      parameters = L"\"" + download_path + L"\" ";
```
Comment (erengy) [inline src/track/feed_aggregator.cpp]: Here you assume that other torrent clients might accept the download path if you just pass it along. They don't. It'd be one thing if they ignored it, but it's worse: qBittorrent, for instance, gives out an error after this change.

Comment (qwe2) [inline src/track/feed_aggregator.cpp]: Yeah but wasn't it the same before with the difference of the "/download" in the argument list? I assumed that was uTorrent specific and would not work for any other client either.

**Comment by qwe2** [comment]
Thanks for the pointers; I'm going to try your branch when I get a chance but looking at the code it does the same thing.

I don't code c++ much and come from a Java/Scala background so over-engineering comes from that I suppose. Seems like I still have a long way to go. Also I always try to write code that's easily extensible but you are right that in this case I overcomplicated it. So thanks again for taking a look at my code.

File: src/track/feed_aggregator.cpp (hunk 2 above)
Comment (erengy) [inline src/track/feed_aggregator.cpp]: Previously, the download path was passed only to uTorrent in the following format:

```bat
utorrent.exe /download "{download_path}" "{torrent_file}"

other.exe "{torrent_file}"
```
  
Here, the download path is passed to all clients:

```bat
utorrent.exe /download "{download_path}" "{torrent_file}"

deluge-console.exe add -p \"{download_path}\" \"{torrent_file}\"

other.exe "{download_path}" "{torrent_file}"
```

qBittorrent tries to open `"{download_path}"` as another torrent file and gives out an error. I imagine that other clients would do the same.

**Comment by qwe2** [comment]
Thanks, your branch is working great.

**Pull request closed without merging**